axonops/axonops-workbench
# #634 Logging

Fixes #634: "Replace custom logger with a standard library."

### What was done

- Removed in-house logger and related files
- Introduced new dependency `electron-log`
- Replaced usage of in-house logger with electron-log
- Added more logs to have better visibility of the app workflows
- Replaced generic log messages (e.g., 'workspaces') with more specific ones ('Failed to save workspace')
- Added logs to the catch blocks that were empty before to avoid error suppression `catch (e) {}` #710
<img width="708" alt="image" src="https://github.com/user-attachments/assets/2c658386-de25-4a8b-911c-3bdd8ea994de" />

- Refactored a few usages of exceptions for the flow control to avoid logging zeroes
- Added more details about logged events, improving clarity and preparing logs for automated processing, if it will be introduced later.

```javascript
log.warning('Failed to move workspace', 
    {'workspace': workspace.name, 'old_path': originalPath, 'new_path': newPath, 'error': e})
```

### How to test

**Due to a lack of automated testing and a big changeset (25 files affected, +570/-1300), we must run extensive manual testing to avoid regression.**

- Fetch the code and switch the branch to [ha/#634-logging](https://github.com/HadesArchitect/axonops-workbench/tree/%23634-logging)
- Execute `npm i` to get the new dependency
- Run the app to see logs in the console. They appear in the DevTools console as well as in the IDE console. I recommend using VSCode with a preconfigured ` main + renderer` debugger.
<img width="727" alt="image" src="https://github.com/user-attachments/assets/3bdc57ee-b270-44d0-9133-01abe4d66045" />

- Review the logs and check if they are correctly describing the app state and processes.
- Enable logging in the configuration and restart the app. A new logfile must appear. The location of the file is OS specific, for OSX it's `/Users/[USER]/Library/Logs/AxonOps Workbench/main.log` Path to the log file is printed in console if the logging to file is enabled.
<img width="465" alt="image" src="https://github.com/user-attachments/assets/bbcfa226-5786-44ed-a6d6-7ee55f691135" />

- Disable logging in the configuration and restart the app. New logs must not appear in the log file. 
- Run through the app features and user scenarios to test if any regression was introduced. (I did it myself but one is not enough)
- Review the changeset. I'm very open for the discussion and any suggestions.

### Special notes
- I've updated the English localization file to specify logging to the file, but I can't do it for other languages.
- As I ran out of available time, some desired features (e.g., configurable LOGLEVEL) haven't been implemented. If those are important/desired, a new ticket must be created.
- For non-info events (errors), LOGLEVEL was chosen based on the error handling in the code. As most errors are suppressed and control flow doesn't change, accordingly a `warning` log-level has been set. Personally, I disagree with many of those cases as I'd consider them as errors, thus `error` log-level would be more suitable, but they aren't processed as errors (control flow goes as if nothing happened). 
```javascript
# main/main.js
  ...
  // Load the main HTML file of that window
  try {
    windowObject.loadURL(URL.format({
      pathname: viewPath,
      protocol: 'file:',
      slashes: true
    }))
  } catch (e) {
    log.warning('Failed to load view file', {'path': viewPath, 'error': e})
    // --->>> TODO: it's an exceptional situation, must be error+abort, not warning <<<---
  }
  
  // Whether or not the window should be at the center of the screen
  if (extraProperties.center)
  ...
```

- Using Exceptions to by-pass parts of code  (`throw 0`) is an anti-pattern. Exceptions are designed for handling exceptional situations, not a regular flow control (https://wiki.c2.com/?DontUseExceptionsForFlowControl).
  - In the workbench, those zeroes are used widely (almost 0.5k cases), so they inevitably will appear in logs
  - As it wasn't the primary goal of the ticket, I mostly ignored it and only refactored a couple of places to demonstrate that the code looks better without them. This screenshot shows how we avoid anti-pattern and decrease line count, making the app less complex -> improving maintainability.
  
<img width="936" alt="image" src="https://github.com/user-attachments/assets/6af68535-c3b3-469b-b35d-bbef11e39e4d" />
